presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -141,6 +146,9 @@
     private final AtomicReference<Optional<Output>> output = new AtomicReference<>(Optional.empty());
     private final StateMachine<Optional<QueryInfo>> finalQueryInfo;
 
+    private final Map<String, String> addedSessionFunctions = new ConcurrentHashMap<>();
+    private final Set<String> removedSessionFunctions = Sets.newConcurrentHashSet();
```

Comment:
I don't like that they're still Strings in `QueryStateMachine` - they should have already been deserialized by now.

Reply:
I've changed everything on the server-side to Map<BasicSignature, SqlInvokedFunction>